### DISCUSSION TITLE
cardSetsInList - How to use?

### DISCUSSION THREAD

**Discussion by Phuzzyhead** [finol-digital/Card-Game-Simulator#74] [Q&A] [answered]
Heyo,

how do I use "cardSetsInList"?
I've seen it in the schema and in the documentation example for grand archive.
But this uses complex object type set definitions.
My case is much simpler, but I can't seem to get it to work.

In AllCards.json I have:
`{"name": "red shirt", "id": "crewT1redshirt", "type": "crew", "set": ["crewT1", "crewT2", "crewT3"], "inStack": "crewT1"}`
In AllSets.json I have:
```
[
    {"code":"crewT1","name":"Crew"},
    {"code":"crewT2","name":"Crew"},
    {"code":"crewT3","name":"Crew"}
]
```
And I have set `"cardSetsInList": true` in my game def.
If set it to `false` and change to `"set": "crewT1"` everything works as expected.
I would expect "cardSetsInList" to duplicate my card for each set it is part of. Is my expectation incorrect? 
Or does it maybe only work together with other options?

Thanks for any help :-)

**Comment by davidmfinol** [comment]
It sounds like you're using it correctly, but maybe the duplicates are being hidden?
In the Settings Menu, there is an option to `Hide Reprints`: https://github.com/finol-digital/Card-Game-Simulator/wiki/Global-Settings
Do the cards appear as you expect if you uncheck `Hide Reprints`?

**Reply by Phuzzyhead** [reply]
Thanks for replying!
I can't check that ATM, because the game does not load. Sorry, should have provided this right away:
```
Error loading game!: Cannot access child value on Newtonsoft.Json.Linq.JValue.  at Newtonsoft.Json.Linq.JToken.get_Item (System.Object key) [0x00000] in <00000000000000000000000000000000>:0 
  at Newtonsoft.Json.Linq.JToken.Value[T] (System.Object key) [0x00000] in <00000000000000000000000000000000>:0 
  at CardGameDef.Unity.UnityCardGame.PopulateCardSets (System.Collections.Generic.Dictionary`2[TKey,TValue] cardSets, Newtonsoft.Json.Linq.JToken cardJToken, System.String defaultSetCode) [0x00000] in <00000000000000000000000000000000>:0 
  at CardGameDef.Unity.UnityCardGame.LoadCardFromJToken (Newtonsoft.Json.Linq.JToken cardJToken, System.String defaultSetCode) [0x00000] in <00000000000000000000000000000000>:0 
  at CardGameDef.Unity.UnityCardGame.LoadJsonFromFile (System.String file, CardGameDef.Unity.LoadJTokenDelegate load, System.String dataId, System.String defaultSetCode) [0x00000] in <00000000000000000000000000000000>:0 
  at CardGameDef.Unity.UnityCardGame.LoadCards (System.Int32 page) [0x00000] in <00000000000000000000000000000000>:0 
  at Cgs.CardGameManager+<LoadCards>d__74.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00000] in <00000000000000000000000000000000>:0 
  at CardGameDef.Unity.UnityCardGame.Load (CardGameDef.Unity.CardGameCoroutineDelegate updateCoroutine, CardGameDef.Unity.CardGameCoroutineDelegate loadCardsCoroutine, CardGameDef.Unity.CardGameCoroutineDelegate loadSetCardsCoroutine) [0x00000] in <00000000000000000000000000000000>:0 
  at Cgs.CardGameManager.ResetGameScene () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.SceneManagement.SceneManager.Internal_SceneLoaded (UnityEngine.SceneManagement.Scene scene, UnityEngine.SceneManagement.LoadSceneMode mode) [0x00000] in <00000000000000000000000000000000>:0 
```
Any Ideas?
Thanks :-)

**Reply by davidmfinol** [reply]
Ah, ok, looks like it was my mistake too. Currently, the entries in `set` need to be complete objects.
So like this:
`{"name": "red shirt", "id": "crewT1redshirt", "type": "crew", "set": [{"code":"crewT1","name":"Crew"}, {"code":"crewT2","name":"Crew"}, {"code":"crewT3","name":"Crew"}], "inStack": "crewT1"}`

If you do this, you can also potentially choose to remove AllSets.json entirely.


----------


EDIT: I am also changing the code in CGS so that it will also work as you expected. It should be in the next release in a couple of weeks.
Basically, I am changing from this:
```
var code = setJToken.Value<string>(dataIdentifier) ?? defaultSetCode;
var name = setJToken.Value<string>(CardSetNameIdentifier) ?? code;
```
to this:
```
var code = setJToken.ToString();
var name = code;
if (setJToken.HasValues)
{
    code = setJToken.Value<string>(dataIdentifier) ?? defaultSetCode;
    name = setJToken.Value<string>(CardSetNameIdentifier) ?? code;
}
```

**Reply by Phuzzyhead** [reply]
Thanks a lot! 
That update will be great to reduce redundant information in AllCards.json.

If the entries need to be complete objects for now, I will also need to set `cardSetIsObject = true`, right?

**Comment by Phuzzyhead** [comment, marked as answer]
Hey there,

I finally got it right.
Here's a recap what I needed in case someone else comes looking:

- Define sets together with cards like this: `"set": [{"code":"crewT1","name":"Crew"}, {"code":"crewT2","name":"Crew"}, {"code":"crewT3","name":"Crew"}]}`
- GameName.json `"cardSetsInList": true`
- GameName.json `"cardSetsInList": true`
- GameName.json add an additional card property:
 `{
      "name": "set",
      "display": "Set",
      "type": "objectList"
    }` 
- GameName.json `"cardSetIdentifier": "set"`

I think there already is a default property named "set", but you can't use that because it doesn't have the right type. The assignment won't work and the set will be the default set. 
Also, making "set" of type "stringList" somehow works with objects but there will be another "Set" property shown ingame without values.

Thanks @davidmfinol for the support. 👍
